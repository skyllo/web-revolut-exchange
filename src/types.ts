export enum Pocket {
  EUR = 'EUR',
  GBP = 'GBP',
  USD = 'USD',
}

export interface PocketData {
  name: Pocket;
  balance: number;
  symbol: string;
  title: string;
  countryCode: string;
}

export type InputDirection = 'from' | 'to';

export interface Rate {
  rate: number;
  fromPocket: Pocket;
  toPocket: Pocket;
}
