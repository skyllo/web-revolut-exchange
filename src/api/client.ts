import { Pocket, PocketData, Rate } from '../types';
import { toFixedDecimals } from '../utils/utils';

interface PostExchangeResult {
  fromBalance: number;
  toBalance: number;
}

export async function postExchange(
  fromPocketData: PocketData,
  toPocketData: PocketData,
  amount: number,
  rate: number
): Promise<PostExchangeResult> {
  const { balance: fromBalance, name: fromPocket } = fromPocketData;
  const { balance: toBalance, name: toPocket } = toPocketData;
  console.log('postExchange', fromPocket, toPocket, amount, rate);
  return {
    fromBalance: toFixedDecimals(fromBalance - amount),
    toBalance: toFixedDecimals(toBalance + amount * rate),
  };
}

export async function getRate(fromPocket: Pocket, toPocket: Pocket): Promise<Rate> {
  console.log('getRate', fromPocket, toPocket);
  const res = await fetch(`https://api.exchangeratesapi.io/latest?base=${fromPocket}`);
  const { rates } = await res.json();
  const rate = toFixedDecimals(Number(rates[toPocket]), true, 4);
  return {
    fromPocket,
    toPocket,
    rate,
  };
}
