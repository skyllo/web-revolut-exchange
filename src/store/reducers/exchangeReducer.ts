/* eslint-disable no-param-reassign */
import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { Pocket } from '../../types';

export interface ExhangeState {
  fromPocket: Pocket;
  toPocket: Pocket;
}

const initialState: ExhangeState = {
  fromPocket: Pocket.GBP,
  toPocket: Pocket.EUR,
};

function switchPockets(state: ExhangeState) {
  const { fromPocket, toPocket } = state;
  state.fromPocket = toPocket;
  state.toPocket = fromPocket;
}

export const exchangeSlice = createSlice({
  name: 'exchange',
  initialState,
  reducers: {
    swapPockets(state) {
      switchPockets(state);
    },
    setFromPocket(state, action: PayloadAction<Pocket>) {
      const { fromPocket, toPocket } = state;
      const { payload: pocket } = action;
      // swap or switch to a new pocket when selected
      if (fromPocket === pocket || toPocket === pocket) {
        switchPockets(state);
      } else {
        state.fromPocket = action.payload;
      }
    },
    setToPocket(state, action: PayloadAction<Pocket>) {
      const { fromPocket, toPocket } = state;
      const { payload: pocket } = action;
      // swap or switch to a new pocket when selected
      if (fromPocket === pocket || toPocket === pocket) {
        switchPockets(state);
      } else {
        state.toPocket = action.payload;
      }
    },
  },
});

export const { setFromPocket, setToPocket, swapPockets } = exchangeSlice.actions;

export default exchangeSlice.reducer;
