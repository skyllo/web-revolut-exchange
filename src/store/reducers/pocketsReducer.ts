/* eslint-disable no-param-reassign */
import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { POCKET_TO_SYMBOL } from '../../utils/constants';
import { Pocket, PocketData } from '../../types';

type PocketAction = PayloadAction<{
  pocket: Pocket;
  balance: number;
}>;

export const initialStatePockets: Record<Pocket, PocketData> = {
  EUR: {
    name: Pocket.EUR,
    balance: 100,
    symbol: POCKET_TO_SYMBOL[Pocket.EUR],
    title: 'Euro',
    countryCode: 'EU',
  },
  GBP: {
    name: Pocket.GBP,
    balance: 200,
    symbol: POCKET_TO_SYMBOL[Pocket.GBP],
    title: 'British Pound',
    countryCode: 'GB',
  },
  USD: {
    name: Pocket.USD,
    balance: 300,
    symbol: POCKET_TO_SYMBOL[Pocket.USD],
    title: 'US Dollar',
    countryCode: 'US',
  },
};

export const pocketsSlice = createSlice({
  name: 'pockets',
  initialState: initialStatePockets,
  reducers: {
    setBalance: (state, { payload: { balance, pocket } }: PocketAction) => {
      state[pocket].balance = balance;
    },
  },
});

export const { setBalance } = pocketsSlice.actions;

export default pocketsSlice.reducer;
