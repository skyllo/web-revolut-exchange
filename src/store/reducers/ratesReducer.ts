/* eslint-disable no-param-reassign */
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import { getRate } from '../../api/client';
import { Pocket, Rate } from '../../types';

interface RateAction {
  fromPocket: Pocket;
  toPocket: Pocket;
}

export interface RateState {
  status: 'loading' | 'succeeded' | 'failed';
  error: string | undefined;
  data: Rate;
}

export const initialStateRate: RateState = {
  status: 'loading',
  error: undefined,
  data: {
    rate: 0,
    fromPocket: Pocket.GBP,
    toPocket: Pocket.EUR,
  },
};

export const fetchRate = createAsyncThunk<Rate, RateAction>(
  'rates/fetchRates',
  async ({ fromPocket, toPocket }) => getRate(fromPocket, toPocket)
);

export const ratesSlice = createSlice({
  name: 'rates',
  initialState: initialStateRate,
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(fetchRate.pending, (state) => {
      state.status = 'loading';
    });
    builder.addCase(fetchRate.fulfilled, (state, action) => {
      state.status = 'succeeded';
      state.data = action.payload;
    });
    builder.addCase(fetchRate.rejected, (state, action) => {
      state.status = 'failed';
      state.error = action.error.message;
    });
  },
});

export default ratesSlice.reducer;
