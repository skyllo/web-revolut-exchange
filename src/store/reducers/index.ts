import { combineReducers } from '@reduxjs/toolkit';
import exchangeReducer from './exchangeReducer';
import pocketsReducer from './pocketsReducer';
import ratesReducer from './ratesReducer';

export default combineReducers({
  exchange: exchangeReducer,
  pockets: pocketsReducer,
  rates: ratesReducer,
});
