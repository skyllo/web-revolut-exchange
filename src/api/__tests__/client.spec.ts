import fetchMock from 'jest-fetch-mock';
import { initialStatePockets } from '../../store/reducers/pocketsReducer';
import { Pocket } from '../../types';
import { getRate, postExchange } from '../client';

beforeAll(() => {
  fetchMock.resetMocks();
});

describe('postExchange()', () => {
  it('should return correct balance after exchange', async () => {
    const newBalances = await postExchange(
      initialStatePockets.GBP,
      initialStatePockets.EUR,
      200,
      1.5
    );
    expect(newBalances).toMatchSnapshot();
  });

  it('should only return 2 decimal places after exchange', async () => {
    const newBalances = await postExchange(
      initialStatePockets.EUR,
      initialStatePockets.USD,
      100,
      1.2345
    );
    expect(newBalances).toMatchSnapshot();
  });
});

describe('getRate()', () => {
  it('should return rate for GBP -> EUR with 4 decimal places', async () => {
    fetchMock.mockOnce(
      JSON.stringify({ rates: { EUR: 1.1162583022, GBP: 1.0 }, base: 'GBP', date: '2020-11-17' })
    );
    const rate = await getRate(Pocket.GBP, Pocket.EUR);
    expect(rate).toMatchSnapshot();
    expect(fetchMock.mock.calls.length).toEqual(1);
    expect(fetchMock.mock.calls[0][0]).toMatchSnapshot();
  });
});
