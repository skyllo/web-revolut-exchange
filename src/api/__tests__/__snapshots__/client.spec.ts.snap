// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getRate() should return rate for GBP -> EUR with 4 decimal places 1`] = `
Object {
  "fromPocket": "GBP",
  "rate": 1.1163,
  "toPocket": "EUR",
}
`;

exports[`getRate() should return rate for GBP -> EUR with 4 decimal places 2`] = `"https://api.exchangeratesapi.io/latest?base=GBP"`;

exports[`postExchange() should only return 2 decimal places after exchange 1`] = `
Object {
  "fromBalance": 0,
  "toBalance": 423.45,
}
`;

exports[`postExchange() should return correct balance after exchange 1`] = `
Object {
  "fromBalance": 0,
  "toBalance": 400,
}
`;
